iocore/net/quic/QUICStreamFactory.cc

```diff
@@ -0,0 +1,98 @@
+/** @file
+ *
+ *  A brief file description
+ *
+ *  @section license License
+ *
+ *  Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "QUICStream.h"
+#include "QUICBidirectionalStream.h"
+#include "QUICUnidirectionalStream.h"
+#include "QUICStreamFactory.h"
+
+ClassAllocator<QUICBidirectionalStream> quicBidiStreamAllocator("quicBidiStreamAllocator");
+ClassAllocator<QUICSendStream> quicSendStreamAllocator("quicSendStreamAllocator");
+ClassAllocator<QUICReceiveStream> quicReceiveStreamAllocator("quicReceiveStreamAllocator");
+
+QUICStreamVConnection *
+QUICStreamFactory::create(QUICStreamId sid, uint64_t local_max_stream_data, uint64_t remote_max_stream_data)
+{
+  QUICStreamVConnection *stream = nullptr;
+  switch (QUICTypeUtil::detect_stream_type(sid)) {
```

Comment:
This switch case looks same logic of `QUICTypeUtil::detect_stream_direction()`. Is there any difference?

Reply:
The `QUICStreamType` is fixed value:
See

Bits | Stream Type
-- | --
0x0 | Client-Initiated, Bidirectional
0x1 | Server-Initiated, Bidirectional
0x2 | Client-Initiated, Unidirectional
0x3 | Server-Initiated, Unidirectional

we can use it and the context to determine which direction should be used.

If we only used `QUICStreamDirection`. We would lost some information about which side we sit (client or server).
```
enum class QUICStreamDirection : uint8_t {
  UNKNOWN = 0,
  SEND,
  RECEIVE,
  BIDIRECTIONAL,
};
```

So QUICStreamDirection indicate which type of stream should be create.
QUICStreamType and context can be used to determine some of logic like transport parameters and stream manager  initialization.

Reply:
I guess @masaori335 was trying to say we could do this:

```
switch(detect_stream_direction(sid, this->_info->direction())) {
case BIDIRECTIONAL:
  ...
case SEND:
  ...
case RECEIVE:
  ...
}
```

Reply:
fixed..

Reply:
Thanks :+1: I thought we would make mistake easily with the complicated logic.